OpenRA.Mods.Common/FileFormats/WavReader.cs

```diff
@@ -99,97 +97,275 @@ public static bool LoadSound(Stream s, out Func<Stream> result, out short channe
+
 				numBlocks = dataSize / blockAlign;
-				blockDataSize = blockAlign - channels * 4;
-				outputSize = uncompressedSize * channels * 2;
-				predictor = new int[channels];
-				index = new int[channels];
+				var numGroupsPerBlock = inBlockDataSize / channels / SamplesPerGroup * BytesPerSample;
 
-				interleaveBuffer = new byte[channels * 16];
+				outGroupSize = SamplesPerGroup * BytesPerSample * channels;
+				var outHead = channels * BytesPerSample;
+				var outGroups = numGroupsPerBlock * outGroupSize;
+				Length = numBlocks * (outHead + outGroups);
+
+				predictor = new short[channels];
+				index = new byte[channels];
 			}
 
-			protected override bool BufferData(Stream baseStream, Queue<byte> data)
+			/// <summary>
+			/// Reads and decodes blocks from the base stream into the ring buffer.
+			/// </summary>
+			bool ReadBlocks()
 			{
-				// Decode each block of IMA ADPCM data
-				// Each block starts with a initial state per-channel
-				for (var c = 0; c < channels; c++)
+				// Check if we have enough space for a group. Ask for a flush if not.
+				// If ring buffer size were too small this could loop forever, however
+				// we know our buffer will always be large enough.
+				var availableBytes = ringBuf.Length - count;
+				if (availableBytes < outGroupSize + 2)
+					return false;
+
+				// PERF: The output is 16-bit PCM, so we can write bytes as if they were shorts for less CPU churn.
+				var outShorts = MemoryMarshal.Cast<byte, short>(ringBuf.AsSpan());
+
+				// PERF: Used for batch reading data from stream.
+				Span<byte> header = stackalloc byte[4];
+				Span<byte> group = stackalloc byte[4 * channels];
+
+				// Fill the ring buffer instead decoding just one group.
+				while (ringBuf.Length - count >= outGroupSize + 2 && currentBlock < numBlocks)
+				{
+					// If we are at the start of a block, read the header
+					if (inBlockOffset == 0)
+					{
+						for (var c = 0; c < channels; c++)
+						{
+							baseStream.ReadExactly(header);
+
+							// Load initial values.
+							predictor[c] = (short)(header[0] | (header[1] << 8));
+							index[c] = header[2];
+
+							// header[3] is unknown/reserved
+
+							// Write initial predictor samples interleaved
+							var pos = tail / 2;
+							outShorts[pos] = predictor[c];
+							tail = (tail + 2) % ringBuf.Length;
+							count += 2;
+						}
+					}
+
+					// Writing into a single buffer to improve cache locality, making sure data is serial.
+					for (var c = 0; c < channels; c++)
+						baseStream.ReadExactly(group.Slice(c * 4, 4));
+
+					// Decode the group sequentially.
+					for (var c = 0; c < channels; c++)
+					{
+						ref var pred = ref predictor[c];
+						ref var idx = ref index[c];
+
+						var offset = c * 4;
+						var b0 = group[offset + 0];
+						var b1 = group[offset + 1];
+						var b2 = group[offset + 2];
+						var b3 = group[offset + 3];
+
+						// Decode into temporary variables so they could be easily turned into registers.
```

Comment:
same as before. doubt this is beneficial, as the compiler may not have 8 registers available anymore - with the already earlier local variables. also after you only use them once.

if you do want to use it then set s0 and after set the value in the array directly.